File: android/liulian-ui/src/main/kotlin/io/liulian/ui/LiulianTab.kt
package io.liulian.ui

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.layout.positionInParent
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import io.liulian.tokens.LiulianTokens

data class LiulianTabItem(
    val id: String,
    val label: String,
    val disabled: Boolean = false,
)

enum class TabVariant { Default, Compact }

/**
 * LiulianTab — segmented horizontal nav per ui-spec/Tab.spec.md.
 *
 * Active indicator is a single bar that animates X position between tabs.
 * That's the LIULIAN signature — NOT a Material per-tab indicator with fade.
 */
@Composable
fun LiulianTab(
    items: List<LiulianTabItem>,
    activeId: String,
    onChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    variant: TabVariant = TabVariant.Default,
) {
    val height = when (variant) {
        TabVariant.Default -> LiulianTokens.Control.Height.md
        TabVariant.Compact -> LiulianTokens.Control.Height.sm
    }
    val itemPaddingX = LiulianTokens.Spacing.s4  // 16
    val density = LocalDensity.current

    // Measured tab positions in raw pixels (Int). Converted to Dp when consumed.
    val positions = remember { mutableStateOf(mapOf<String, Pair<Int, Int>>()) }

    val activePos = positions.value[activeId]
    val indicatorX by animateDpAsState(
        targetValue = with(density) { (activePos?.first ?: 0).toDp() },
        animationSpec = tween(
            durationMillis = LiulianTokens.Duration.fast,
            easing = LiulianTokens.Easing.easeOutQuart,
        ),
        label = "tabIndicatorX",
    )
    val indicatorWidth by animateDpAsState(
        targetValue = with(density) { (activePos?.second ?: 0).toDp() },
        animationSpec = tween(
            durationMillis = LiulianTokens.Duration.fast,
            easing = LiulianTokens.Easing.easeOutQuart,
        ),
        label = "tabIndicatorWidth",
    )

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .background(LiulianTokens.Colors.canvasWarm)
            .border(width = 0.dp, color = LiulianTokens.Colors.hairline),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            items.forEach { item ->
                val isActive = item.id == activeId
                val source = remember { MutableInteractionSource() }
                Box(
                    modifier = Modifier
                        .height(height)
                        .alpha(if (item.disabled) LiulianTokens.Opacity.disabled else 1f)
                        .clickable(
                            interactionSource = source,
                            indication = null,
                            enabled = !item.disabled,
                            onClick = { onChange(item.id) },
                        )
                        .padding(horizontal = itemPaddingX)
                        .onGloballyPositioned { coords ->
                            val xPx = coords.positionInParent().x.toInt()
                            val wPx = coords.size.width
                            positions.value = positions.value + (item.id to (xPx to wPx))
                        },
                    contentAlignment = Alignment.Center,
                ) {
                    LiulianText(
                        text = item.label,
                        variant = LiulianTextVariant.MonoLabel,
                        color = if (isActive) LiulianTokens.Colors.inkCharcoal else LiulianTokens.Colors.inkMuted,
                    )
                }
            }
        }

        // Active indicator + hairline bottom border, drawn in the same Box
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(1.dp)
                .offset(y = (height.value - 1).dp)
                .background(LiulianTokens.Colors.hairline),
        )
        if (activePos != null) {
            Box(
                modifier = Modifier
                    .offset(x = indicatorX, y = (height.value - 2).dp)
                    .height(2.dp)
                    .width(indicatorWidth)
                    .background(LiulianTokens.Colors.unibeRed),
            )
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFBFBFA)
@Composable
private fun LiulianTabPreview() {
    var active by remember { mutableStateOf("forecast") }
    LiulianTab(
        items = listOf(
            LiulianTabItem("landing", "LANDING"),
            LiulianTabItem("forecast", "FORECAST"),
            LiulianTabItem("studio", "STUDIO"),
        ),
        activeId = active,
        onChange = { active = it },
    )
}

File: ios/LiulianUI/Sources/LiulianUI/LiulianTab.swift
import SwiftUI

/// LiulianTab — per ui-spec/Tab.spec.md.
///
/// Active indicator is a single 2pt bar that animates X between tab positions.
/// LIULIAN signature — NOT a UISegmentedControl pill, NOT a Material per-tab indicator.
public struct LiulianTabItem: Identifiable {
    public let id: String
    public let label: String
    public let disabled: Bool

    public init(id: String, label: String, disabled: Bool = false) {
        self.id = id
        self.label = label
        self.disabled = disabled
    }
}

public enum LiulianTabVariant {
    case `default`, compact
}

private struct TabFrame: Equatable {
    let id: String
    let x: CGFloat
    let width: CGFloat
}

private struct TabFramesKey: PreferenceKey {
    static var defaultValue: [TabFrame] = []
    static func reduce(value: inout [TabFrame], nextValue: () -> [TabFrame]) {
        value.append(contentsOf: nextValue())
    }
}

public struct LiulianTab: View {
    let items: [LiulianTabItem]
    let activeId: String
    let onChange: (String) -> Void
    let variant: LiulianTabVariant

    @State private var frames: [TabFrame] = []

    public init(
        items: [LiulianTabItem],
        activeId: String,
        onChange: @escaping (String) -> Void,
        variant: LiulianTabVariant = .default
    ) {
        self.items = items
        self.activeId = activeId
        self.onChange = onChange
        self.variant = variant
    }

    private var height: CGFloat {
        variant == .compact ? LiulianTokens.Control.Height.sm : LiulianTokens.Control.Height.md
    }

    public var body: some View {
        ZStack(alignment: .bottomLeading) {
            HStack(spacing: 0) {
                ForEach(items) { item in
                    Button(action: {
                        if !item.disabled { onChange(item.id) }
                    }) {
                        LiulianText(
                            item.label,
                            variant: .monoLabel,
                            color: item.id == activeId ? LiulianTokens.Colors.inkCharcoal : LiulianTokens.Colors.inkMuted
                        )
                        .padding(.horizontal, LiulianTokens.Spacing.s4)
                        .frame(height: height)
                        .background(
                            GeometryReader { geo in
                                Color.clear.preference(
                                    key: TabFramesKey.self,
                                    value: [TabFrame(id: item.id, x: geo.frame(in: .named("tabContainer")).minX, width: geo.size.width)]
                                )
                            }
                        )
                    }
                    .buttonStyle(.plain)
                    .opacity(item.disabled ? LiulianTokens.Opacity.disabled : 1.0)
                }
            }
            // hairline bottom
            Rectangle()
                .fill(LiulianTokens.Colors.hairline)
                .frame(height: 1)

            // active indicator
            if let active = frames.first(where: { $0.id == activeId }) {
                Rectangle()
                    .fill(LiulianTokens.Colors.unibeRed)
                    .frame(width: active.width, height: 2)
                    .offset(x: active.x)
                    .animation(
                        .timingCurve(0.16, 1.0, 0.3, 1.0, duration: LiulianTokens.Duration.fast),
                        value: activeId
                    )
            }
        }
        .frame(height: height)
        .background(LiulianTokens.Colors.canvasWarm)
        .coordinateSpace(name: "tabContainer")
        .onPreferenceChange(TabFramesKey.self) { newFrames in
            frames = newFrames
        }
    }
}

#if DEBUG
struct LiulianTab_Previews: PreviewProvider {
    struct PreviewContainer: View {
        @State var active = "forecast"
        var body: some View {
            LiulianTab(
                items: [
                    .init(id: "landing", label: "LANDING"),
                    .init(id: "forecast", label: "FORECAST"),
                    .init(id: "studio", label: "STUDIO"),
                ],
                activeId: active,
                onChange: { active = $0 }
            )
        }
    }
    static var previews: some View {
        PreviewContainer()
    }
}
#endif
